### PR TITLE
[SDAG] Improve `SimplifyDemandedBits` for mul

### LINKED ISSUE #56645: Possible improvement for SimplifyDemandedBits for Mul
If the RHS is a constant with X trailing zeros, then the X MSBs of the LHS are not demanded.

I don't have any plans to do anything with this right now. Just something I thought of while working on https://reviews.llvm.org/D130146 That patch does special isel for (i64 (mul (and X, 0xffffffff), 3 << C)) and requires C < 32. If C >= 32 the AND could be removed, but it doesn't happen today.

FYI @spatel-gh @RKSimon

### PR DESCRIPTION
**Pull request by dtcxzyw**
If the RHS is a constant with X trailing zeros, then the X MSBs of the LHS are not demanded.

Alive2: https://alive2.llvm.org/ce/z/F5CyJW
Fixes https://github.com/llvm/llvm-project/issues/56645.

### CODE REVIEW

**Comment by dtcxzyw** [comment]
> Failed Tests (1):
  LLVM :: CodeGen/Thumb2/mve-vecreduce-add.ll

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 1]
```diff
@@ -2782,10 +2782,17 @@ bool TargetLowering::SimplifyDemandedBits(
     unsigned DemandedBitsLZ = DemandedBits.countl_zero();
     APInt LoMask = APInt::getLowBitsSet(BitWidth, BitWidth - DemandedBitsLZ);
     KnownBits KnownOp0, KnownOp1;
-    if (SimplifyDemandedBits(Op0, LoMask, DemandedElts, KnownOp0, TLO,
-                             Depth + 1) ||
-        SimplifyDemandedBits(Op1, LoMask, DemandedElts, KnownOp1, TLO,
+    auto GetDemandedBitsLHSMask = [&](APInt Demanded,
+                                      const KnownBits &KnownRHS) {
+      if (Op.getOpcode() == ISD::MUL)
+        Demanded &= APInt::getLowBitsSet(
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp]: Done. If necessary, I can split the APInt change into an NFC patch.

**Review by RKSimon — COMMENTED** [review]
The APInt::clearHighBits part LGTM if you want to pull it out into its own PR and I'll accept it

**Comment by dtcxzyw** [comment]
> The APInt::clearHighBits part LGTM if you want to pull it out into its own PR and I'll accept it

See https://github.com/llvm/llvm-project/pull/91938

**Comment by dtcxzyw** [comment]
Rebased.

**Comment by goldsteinn** [comment]
LGTM, wait one 1 more.

**Comment by dtcxzyw** [comment]
Ping.

**Comment by goldsteinn** [comment]
> Ping.

If you don't get any feedback in the next few days, I think you can push w/ my approval.

**Review by RKSimon — APPROVED** [review]
LGTM

**Pull request merged**